codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,210 @@
+
+	index = make(map[protowire.Number]*tagAndProtoType)
+	for i, n := 0, typ.NumField(); i < n; i++ {
+		field := typ.Field(i)
+		protoTag, ok := field.Tag.Lookup("protobuf")
+		if !ok {
+			// Try if it is a protobuf_oneof
+			protoTag, ok = field.Tag.Lookup("protobuf_oneof")
+			if !ok {
+				continue
+			}
+		}
+		byCommaSplits := strings.Split(protoTag, ",")
+		if len(byCommaSplits) == 0 {
+			// Strange, how can a protobuf tag not be populated?
+			continue
+		}
+		wireType, protoFieldNumStr := byCommaSplits[0], byCommaSplits[1]
+		protoField, err := strconv.ParseUint(protoFieldNumStr, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("invalid protoFieldNum in %q", protoFieldNumStr)
+		}
+		tagNum := protowire.Number(protoField)
+
+		wireTypeType, ok := wireTypeStrToWireType[wireType]
+		if !ok {
+			panic(fmt.Sprintf("unhandled wireType %q", wireType))
+		}
+		index[tagNum] = &tagAndProtoType{
+			tagNum:    tagNum,
+			protoType: wireType,
+			wireType:  wireTypeType,
+		}
+	}
+
+	// By this point we should memoize the index.
+	typeToIndexFieldsForType[typ] = index
+
+	return index, nil
+}
+
+var wireTypeStrToWireType = map[string]protowire.Type{
+	"varint":      0,
+	"fixed64":     1,
+	"bytes":       2,
+	"start_group": 3,
+	"end_group":   4,
+	"fixed32":     5,
+}
+
+func wireTypeToString(wt protowire.Type) string {
+	switch wt {
+	case 0:
+		return "varint"
+	case 1:
+		return "fixed64"
+	case 2:
+		return "bytes"
+	case 3:
+		return "start_group"
+	case 4:
+		return "end_group"
+	case 5:
+		return "fixed32"
+	default:
+		return fmt.Sprintf("Unknown type: %d", wt)
+	}
+}
+
+type MismatchedWireType struct {
+	GotWireType  protowire.Type
+	WantWireType protowire.Type
+	TagNum       protowire.Number
+}
+
+func (mwt *MismatchedWireType) String() string {
+	return fmt.Sprintf("Mismatched{TagNum: %d, GotWireType: %q != WantWireType: %q}",
+		mwt.TagNum, wireTypeToString(mwt.GotWireType), wireTypeToString(mwt.WantWireType))
+}
+
+func (mwt *MismatchedWireType) Error() string {
+	return mwt.String()
+}
+
+var _ error = (*MismatchedWireType)(nil)
+
+// CheckMismatchedProtoFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the extraneous fields by tagNumber and wireType, as well as mismatched
+// wireTypes.
+func CheckMismatchedProtoFields(b []byte, msg gogoproto.Message) error {
+	if msg == nil {
+		return errors.New("expecting a non-nil message")
+	}
+	rv := reflect.ValueOf(msg)
+	rt := rv.Elem().Type()
+
+	index, err := indexFieldsForType(rt)
+	if err != nil {
+		return err
+	}
+
+	mismatchedFields := make([]error, 0, len(index))
+
+	for len(b) > 0 {
+		tagNum, wireType, n := protowire.ConsumeField(b)
+		if n < 0 {
+			return errors.New("tagLen is invalid")
+		}
+
+		twt, ok := index[tagNum]
+		switch {
+		case ok:
+			// Now assert that the wireTypes match.
+			if twt.wireType != wireType {
```

Comment:
I'm not sure we really care about this case... maybe it's good to check for it, but it will get checked elsewhere in the actual parsing...

Reply:
Yeah initially in the first PR, I had delegated it to the actual parsing but in tests, it became more apparent that what we are targeting: aka detecting outdated services/messages, malicious behavior that without it, folks can send mismatched types but sharing the same field number.